File: src/service/umami.service.ts
import { SSqqController } from 'src/controller/ssqq.controller';

export class UmamiService {
    private static address = 'https://status.stapxs.cn/api';

    private ready = false;
    private token: string | undefined;

    constructor() {
        this.token = undefined;
        this.getToken().then(() => {
            this.token = this.token;
            this.ready = true;
        });
    }

    public async getUmamiPageviews(time: string, unit: string, query: { [key: string]: any }) {
        if(!time) {
            // 缺失默认为最近 24 小时
            time = `${Date.now() - 86400000}-${Date.now()}`;
        }
        if(time.indexOf('-') === -1 || !/^\d{13}$/.test(time.split('-')[0]) || !/^\d{13}$/.test(time.split('-')[1])) {
            return { error: '参数 time 格式错误，正确格式为 start-end（时间戳，毫秒）' };
        }

        if(SSqqController.removeName.includes(query.type) || query.unit) {
            return { error: '请求参数中包含不被允许的字段' };
        }

        const data = await this.getData(`/websites/${process.env.UMAMI_SITE_ID}/pageviews`, {
            startAt: Number(time.split('-')[0]),
            endAt: Number(time.split('-')[1]),
            unit: unit,
            ...query
        });
        return data;
    }

    public async getUmamiStatus(time: string, query: { [key: string]: any }) {
        if(!time) {
            // 缺失默认为最近 24 小时
            time = `${Date.now() - 86400000}-${Date.now()}`;
        }
        if(time.indexOf('-') === -1 || !/^\d{13}$/.test(time.split('-')[0]) || !/^\d{13}$/.test(time.split('-')[1])) {
            return { error: '参数 time 格式错误，正确格式为 start-end（时间戳，毫秒）' };
        }

        if(SSqqController.removeName.includes(query.type)) {
            return { error: '请求参数中包含不被允许的字段' };
        }

        const data = await this.getData(`/websites/${process.env.UMAMI_SITE_ID}/stats`, {
            startAt: Number(time.split('-')[0]),
            endAt: Number(time.split('-')[1]),
            ...query
        });
        return data;
    }

    public async getUmamiMetrics(name: string, time: string, query: { [key: string]: any }) {
        if(!time) {
            // 缺失默认为最近 24 小时
            time = `${Date.now() - 86400000}-${Date.now()}`;
        }
        if(time.indexOf('-') === -1 || !/^\d{13}$/.test(time.split('-')[0]) || !/^\d{13}$/.test(time.split('-')[1])) {
            return { error: '参数 time 格式错误，正确格式为 start-end（时间戳，毫秒）' };
        }

        if(SSqqController.removeName.includes(name) || name === 'host') {
            return { error: `请求的数据类型 ${name} 不被允许` };
        }

        if(SSqqController.removeName.includes(name) || query.type) {
            return { error: '请求参数中包含不被允许的字段' };
        }

        const data = await this.getData(`/websites/${process.env.UMAMI_SITE_ID}/metrics`, {
            startAt: Number(time.split('-')[0]),
            endAt: Number(time.split('-')[1]),
            type: name,
            ...query
        });
        return data;
    }

    public async getUmamiEvents(time: string) {
        if(!time) {
            // 缺失默认为最近 24 小时
            time = `${Date.now() - 86400000}-${Date.now()}`;
        }
        if(time.indexOf('-') <= 0 || !/^\d{13}$/.test(time.split('-')[0]) || !/^\d{13}$/.test(time.split('-')[1])) {
            return { error: '参数 time 格式错误，正确格式为 start-end（时间戳，毫秒）' };
        }

        const data = await this.getData(`/websites/${process.env.UMAMI_SITE_ID}/event-data/events`, {
            startAt: Number(time.split('-')[0]),
            endAt: Number(time.split('-')[1])
        }) as {
            eventName: string;
            propertyName: string;
            dataType: string;
            total: number;
        }[];

        if(data instanceof Object && (data as any).error) {
            return data;
        }

        // 想要排除的 eventName 列表
        const excludedEvents = ['link_view'];
    
        const filteredData = data.filter(
            (event) => !excludedEvents.includes(event.eventName)
        );

        // 获取事件具体数量
        const detailedDataPromises = data.map(async (event) => {
            const detailedData = await this.getData(`/websites/${process.env.UMAMI_SITE_ID}/event-data/values`, {
                startAt: Number(time.split('-')[0]),
                endAt: Number(time.split('-')[1]),
                eventName: event.eventName,
                propertyName: event.propertyName
            });
            return {
                ...event,
                details: detailedData
            };
        });

        const detailedData = await Promise.all(detailedDataPromises);
        return detailedData;
    }

    public async getUmamiSessions(time: string) {
        if(!time) {
            // 缺失默认为最近 24 小时
            time = `${Date.now() - 86400000}-${Date.now()}`;
        }
        if(time.indexOf('-') <= 0 || !/^\d{13}$/.test(time.split('-')[0]) || !/^\d{13}$/.test(time.split('-')[1])) {
            return { error: '参数 time 格式错误，正确格式为 start-end（时间戳，毫秒）' };
        }

        const data = await this.getData(`/websites/${process.env.UMAMI_SITE_ID}/session-data/properties`, {
            startAt: Number(time.split('-')[0]),
            endAt: Number(time.split('-')[1])
        }) as {
            propertyName: string;
            total: number;
        }[];

        if(data instanceof Object && (data as any).error) {
            return data;
        }

        const detailedDataPromises = data.map(async (item) => {
            const detailedData = await this.getData(`/websites/${process.env.UMAMI_SITE_ID}/session-data/values`, {
                startAt: Number(time.split('-')[0]),
                endAt: Number(time.split('-')[1]),
                propertyName: item.propertyName
            });
            return {
                ...item,
                details: detailedData
            };
        });

        const detailedData = await Promise.all(detailedDataPromises);
        return detailedData;
    }

    // ==================================================================

    /**
     * 获取 Umami API 的访问 token
     * @returns Umami API 的访问 token
     */
    private async getToken(): Promise<string | undefined> {
        const path = '/auth/login';
        const body = {
            username: process.env.UMAMI_USER,
            password: process.env.UMAMI_PASSWORD
        };
        const res = await fetch(UmamiService.address + path, {
            method: 'POST',
            headers: {
                'content-type': 'application/json'
            },
            body: JSON.stringify(body)
        });
        const result = await res.json();
        if (result.token) {
            return result.token;
        } else {
            throw new Error('Umami 登录失败');
        }
    }

    /**
     * 检查当前 token 是否有效
     * @returns Token 是否有效
     */
    private async checkToken(): Promise<boolean> {
        const path = '/auth/verify';
        const res = await fetch(UmamiService.address + path, {
            method: 'POST',
            headers: {
                'content-type': 'application/json',
                'Authorization': `Bearer ${this.token}`
            },
        });
        const result = await res.text();
        try {
            const data = JSON.parse(result);
            if(data.id != undefined) {
                return true;
            }
        } catch (e) {
            // do nothing
        }
        return false;
    }

    /**
     * 获取 Umami API 的数据(GET 请求)
     * @param path API 路径
     * @param params 查询参数
     * @returns Umami API 的响应数据
     */
    public async getData(path: string, params: any = {}): Promise<any> {
        if(!this.ready) {
            return { error: 'Umami 服务尚未准备好，请稍后再试' };
        }

        if (!this.token) {
            this.token = await this.getToken();
        } else {
            const valid = await this.checkToken();
            if (!valid) {
                this.token = await this.getToken();
            }
        }
        const url = new URL(UmamiService.address + path);
        Object.keys(params).forEach(key => url.searchParams.append(key, params[key]));
        const res = await fetch(url.toString(), {
            method: 'GET',
            headers: {
                'content-type': 'application/json',
                'Authorization': `Bearer ${this.token}`
            },
        });
        const result = await res.text();
        try {
            return JSON.parse(result);
        } catch (e) {
            return {
                error: result
            }
        }
    }
}

File: internal/service/umami_service.go
package service

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"sync"
	"time"
)

type UmamiService struct {
	address   string
	siteID    string
	user      string
	password  string
	http      *http.Client
	token     string
	tokenLock sync.Mutex
}

func NewUmamiService(siteID, user, password string) *UmamiService {
	return &UmamiService{
		address:  "https://status.stapxs.cn/api",
		siteID:   siteID,
		user:     user,
		password: password,
		http: &http.Client{
			Timeout: 12 * time.Second,
		},
	}
}

func (s *UmamiService) GetActive() any {
	return s.getData(fmt.Sprintf("/websites/%s/active", s.siteID), map[string]any{})
}

func (s *UmamiService) GetPageviews(timeRange, unit string, query map[string]any) any {
	timeRange = ensureTimeRange(timeRange)
	if !isValidTimeRange(timeRange) {
		return map[string]any{"error": "参数 time 格式错误，正确格式为 start-end（时间戳，毫秒）"}
	}
	if containsForbiddenType(query, "country", "region", "city") || query["unit"] != nil {
		return map[string]any{"error": "请求参数中包含不被允许的字段"}
	}
	startAt, endAt := splitTimeRange(timeRange)
	params := map[string]any{"startAt": startAt, "endAt": endAt, "unit": unit}
	mergeMap(params, query)
	return s.getData(fmt.Sprintf("/websites/%s/pageviews", s.siteID), params)
}

func (s *UmamiService) GetStatus(timeRange string, query map[string]any) any {
	timeRange = ensureTimeRange(timeRange)
	if !isValidTimeRange(timeRange) {
		return map[string]any{"error": "参数 time 格式错误，正确格式为 start-end（时间戳，毫秒）"}
	}
	if containsForbiddenType(query, "country", "region", "city") {
		return map[string]any{"error": "请求参数中包含不被允许的字段"}
	}
	startAt, endAt := splitTimeRange(timeRange)
	params := map[string]any{"startAt": startAt, "endAt": endAt}
	mergeMap(params, query)
	return s.getData(fmt.Sprintf("/websites/%s/stats", s.siteID), params)
}

func (s *UmamiService) GetMetrics(name, timeRange string, query map[string]any) any {
	timeRange = ensureTimeRange(timeRange)
	if !isValidTimeRange(timeRange) {
		return map[string]any{"error": "参数 time 格式错误，正确格式为 start-end（时间戳，毫秒）"}
	}
	if name == "host" || inStringSet(name, "country", "region", "city") {
		return map[string]any{"error": fmt.Sprintf("请求的数据类型 %s 不被允许", name)}
	}
	if query["type"] != nil {
		return map[string]any{"error": "请求参数中包含不被允许的字段"}
	}
	startAt, endAt := splitTimeRange(timeRange)
	params := map[string]any{"startAt": startAt, "endAt": endAt, "type": name}
	mergeMap(params, query)
	return s.getData(fmt.Sprintf("/websites/%s/metrics", s.siteID), params)
}

func (s *UmamiService) GetEvents(timeRange string) any {
	timeRange = ensureTimeRange(timeRange)
	if !isValidTimeRange(timeRange) {
		return map[string]any{"error": "参数 time 格式错误，正确格式为 start-end（时间戳，毫秒）"}
	}
	startAt, endAt := splitTimeRange(timeRange)
	baseParams := map[string]any{"startAt": startAt, "endAt": endAt}
	data := s.getData(fmt.Sprintf("/websites/%s/event-data/properties", s.siteID), baseParams)

	items, ok := data.([]any)
	if !ok {
		return data
	}

	result := make([]map[string]any, 0, len(items))
	for _, item := range items {
		event, ok := item.(map[string]any)
		if !ok {
			continue
		}
		eventName, _ := event["eventName"].(string)
		if eventName == "link_view" {
			continue
		}
		propertyName, _ := event["propertyName"].(string)
		detailParams := map[string]any{"startAt": startAt, "endAt": endAt, "eventName": eventName, "propertyName": propertyName}
		event["details"] = s.getData(fmt.Sprintf("/websites/%s/event-data/values", s.siteID), detailParams)
		result = append(result, event)
	}
	return result
}

func (s *UmamiService) GetSessions(timeRange string, filter map[string]any) any {
	timeRange = ensureTimeRange(timeRange)
	if !isValidTimeRange(timeRange) {
		return map[string]any{"error": "参数 time 格式错误，正确格式为 start-end（时间戳，毫秒）"}
	}
	startAt, endAt := splitTimeRange(timeRange)
	baseParams := map[string]any{"startAt": startAt, "endAt": endAt}
	mergeMap(baseParams, filter)
	data := s.getData(fmt.Sprintf("/websites/%s/session-data/properties", s.siteID), baseParams)

	items, ok := data.([]any)
	if !ok {
		return data
	}
	result := make([]map[string]any, 0, len(items))
	for _, item := range items {
		session, ok := item.(map[string]any)
		if !ok {
			continue
		}
		propertyName, _ := session["propertyName"].(string)
		detailParams := map[string]any{"startAt": startAt, "endAt": endAt, "propertyName": propertyName}
		mergeMap(detailParams, filter)
		session["details"] = s.getData(fmt.Sprintf("/websites/%s/session-data/values", s.siteID), detailParams)
		result = append(result, session)
	}
	return result
}

func (s *UmamiService) getData(path string, params map[string]any) any {
	if strings.TrimSpace(s.siteID) == "" || strings.TrimSpace(s.user) == "" || strings.TrimSpace(s.password) == "" {
		return map[string]any{"error": "Umami 配置缺失"}
	}

	token, err := s.ensureToken()
	if err != nil {
		return map[string]any{"error": err.Error()}
	}

	u, _ := url.Parse(s.address + path)
	q := u.Query()
	for k, v := range params {
		q.Set(k, fmt.Sprintf("%v", v))
	}
	u.RawQuery = q.Encode()

	req, _ := http.NewRequest(http.MethodGet, u.String(), nil)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+token)

	resp, err := s.http.Do(req)
	if err != nil {
		return map[string]any{"error": err.Error()}
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)

	var out any
	if err := json.Unmarshal(body, &out); err != nil {
		return map[string]any{"error": string(body)}
	}
	return out
}

func (s *UmamiService) ensureToken() (string, error) {
	s.tokenLock.Lock()
	defer s.tokenLock.Unlock()

	if s.token != "" {
		ok, _ := s.checkToken(s.token)
		if ok {
			return s.token, nil
		}
	}

	token, err := s.getToken()
	if err != nil {
		return "", err
	}
	s.token = token
	return token, nil
}

func (s *UmamiService) getToken() (string, error) {
	payload := map[string]string{"username": s.user, "password": s.password}
	buf, _ := json.Marshal(payload)
	req, _ := http.NewRequest(http.MethodPost, s.address+"/auth/login", bytes.NewReader(buf))
	req.Header.Set("Content-Type", "application/json")
	resp, err := s.http.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)

	var out map[string]any
	if err := json.Unmarshal(body, &out); err != nil {
		return "", fmt.Errorf("Umami 登录失败")
	}
	token, _ := out["token"].(string)
	if token == "" {
		return "", fmt.Errorf("Umami 登录失败")
	}
	return token, nil
}

func (s *UmamiService) checkToken(token string) (bool, error) {
	req, _ := http.NewRequest(http.MethodPost, s.address+"/auth/verify", nil)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+token)
	resp, err := s.http.Do(req)
	if err != nil {
		return false, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)

	var out map[string]any
	if err := json.Unmarshal(body, &out); err != nil {
		return false, nil
	}
	_, exists := out["id"]
	return exists, nil
}

func ensureTimeRange(timeRange string) string {
	if strings.TrimSpace(timeRange) != "" {
		return timeRange
	}
	end := time.Now().UnixMilli()
	start := end - 86400000
	return fmt.Sprintf("%d-%d", start, end)
}

func isValidTimeRange(timeRange string) bool {
	parts := strings.Split(timeRange, "-")
	if len(parts) != 2 {
		return false
	}
	r := regexp.MustCompile(`^\d{13}$`)
	return r.MatchString(parts[0]) && r.MatchString(parts[1])
}

func splitTimeRange(timeRange string) (string, string) {
	parts := strings.Split(timeRange, "-")
	return parts[0], parts[1]
}

func mergeMap(dst map[string]any, src map[string]any) {
	for k, v := range src {
		dst[k] = v
	}
}

func inStringSet(target string, list ...string) bool {
	for _, item := range list {
		if target == item {
			return true
		}
	}
	return false
}

func containsForbiddenType(query map[string]any, forbidden ...string) bool {
	v, exists := query["type"]
	if !exists {
		return false
	}
	value := fmt.Sprintf("%v", v)
	for _, item := range forbidden {
		if value == item {
			return true
		}
	}
	return false
}
